File: src/main/java/net/katsstuff/danmakucore/entity/spellcard/spellcardbar/SpellcardInfoServer.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.entity.spellcard.spellcardbar;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import net.katsstuff.danmakucore.network.DanmakuCorePacketHandler;
import net.katsstuff.danmakucore.network.SpellcardInfoPacket;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.text.ITextComponent;

public class SpellcardInfoServer extends SpellcardInfo {

	private float position = 1F;
	private final float destinationHeight = 0.02F;

	private List<EntityPlayerMP> sentTo = new ArrayList<>();

	public SpellcardInfoServer(UUID uuid, ITextComponent name, boolean mirrorText) {
		super(uuid, name, mirrorText);
	}

	public SpellcardInfoServer(ITextComponent name) {
		this(MathHelper.getRandomUUID(), name, true);
	}

	public void setNoAnimation() {
		position = destinationHeight;
		sendPacketAll(SpellcardInfoPacket.Action.SET_POS_ABSOLUTE);
	}

	public void tick() {
		if(position > destinationHeight) {
			position = (float)(position - 0.35F * getAcceleration());

			if(position < destinationHeight) {
				position = destinationHeight;
				sendPacketAll(SpellcardInfoPacket.Action.SET_POS_ABSOLUTE);
			}
			else {
				sendPacketAll(SpellcardInfoPacket.Action.SET_POS);
			}
		}
	}

	@Override
	public void setName(ITextComponent name) {
		super.setName(name);
		sendPacketAll(SpellcardInfoPacket.Action.SET_NAME);
	}

	private double getAcceleration() {
		return (position - 1F) * -1 + 0.01;
	}

	@Override
	public void setMirrorText(boolean mirrorText) {
		super.setMirrorText(mirrorText);
		sendPacketAll(SpellcardInfoPacket.Action.SET_MIRROR);
	}

	@Override
	public float getPosX() {
		return 1F;
	}

	@Override
	public float getPosY() {
		return position;
	}

	public void addPlayer(EntityPlayerMP player) {
		sentTo.add(player);
		sendPacket(SpellcardInfoPacket.Action.ADD, player);
	}

	public void removePlayer(EntityPlayerMP player) {
		sentTo.remove(player);
		sendPacket(SpellcardInfoPacket.Action.REMOVE, player);
	}

	public void clear() {
		sendPacketAll(SpellcardInfoPacket.Action.REMOVE);
		sentTo.clear();
	}

	private void sendPacket(SpellcardInfoPacket.Action action, EntityPlayerMP player) {
		DanmakuCorePacketHandler.INSTANCE.sendTo(new SpellcardInfoPacket.Message(this, action), player);
	}

	private void sendPacketAll(SpellcardInfoPacket.Action action) {
		SpellcardInfoPacket.Message message = new SpellcardInfoPacket.Message(this, action);
		sentTo.forEach(p -> DanmakuCorePacketHandler.INSTANCE.sendTo(message, p));
	}
}

File: src/main/scala/net/katsstuff/teamnightclipse/danmakucore/entity/spellcard/spellcardbar/SpellcardInfoServer.scala
/*
 * Copyright (C) 2018  Katrix
 * This file is part of DanmakuCore.
 *
 * DanmakuCore is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DanmakuCore is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with DanmakuCore.  If not, see <https://www.gnu.org/licenses/>.
 */
package net.katsstuff.teamnightclipse.danmakucore.entity.spellcard.spellcardbar

import java.util.UUID

import scala.collection.mutable.ArrayBuffer

import net.katsstuff.teamnightclipse.danmakucore.network._
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.util.math.MathHelper
import net.minecraft.util.text.ITextComponent

class SpellcardInfoServer(uuid: UUID, name: ITextComponent, mirrorText: Boolean)
    extends SpellcardInfo(uuid, name, mirrorText) {
  final val DestinationHeight = 0.02F

  private var position = 1F
  private val sentTo   = new ArrayBuffer[EntityPlayerMP]

  def this(name: ITextComponent) {
    this(MathHelper.getRandomUUID, name, true)
  }

  def setNoAnimation(): Unit = {
    position = DestinationHeight
    sendPacketAll(SetSpellcardInfoPos(uuid, getPosX, getPosY, absolute = true))
  }

  def tick(): Unit = if (position > DestinationHeight) {
    position = (position - 0.35F * getAcceleration).toFloat
    if (position < DestinationHeight) {
      position = DestinationHeight
      sendPacketAll(SetSpellcardInfoPos(uuid, getPosX, getPosY, absolute = true))
    } else sendPacketAll(SetSpellcardInfoPos(uuid, getPosX, getPosY, absolute = false))
  }

  override def setName(name: ITextComponent): Unit = {
    super.setName(name)
    sendPacketAll(SetSpellcardInfoName(uuid, name))
  }

  private def getAcceleration = (position - 1F) * -1 + 0.01
  override def setMirrorText(mirrorText: Boolean): Unit = {
    super.setMirrorText(mirrorText)
    sendPacketAll(SetSpellcardInfoMirror(uuid, mirrorText))
  }

  override def getPosX        = 1F
  override def getPosY: Float = position
  def addPlayer(player: EntityPlayerMP): Unit = {
    sentTo += player
    DanCorePacketHandler.sendTo(AddSpellcardInfo(uuid, name, mirrorText, getPosX, getPosY), player)
  }

  def removePlayer(player: EntityPlayerMP): Unit = {
    sentTo -= player
    DanCorePacketHandler.sendTo(RemoveSpellcardInfo(uuid), player)
  }

  def clear(): Unit = {
    sendPacketAll(RemoveSpellcardInfo(uuid))
    sentTo.clear()
  }

  private def sendPacketAll(packet: SpellcardInfoPacket): Unit =
    sentTo.foreach(p => DanCorePacketHandler.sendTo(packet, p))
}
